Speed up "VMs & Instances in My LDAP Group" filter in /vm_or_template/explorer

NickLaMuro opened the pull request:
Purpose or Intent
-----------------
Speeds up "VMs & Instances in My LDAP Group" index and ajax calls in the `/vm_or_template` (Services -> Workload) section of the application.


How
---
This adds `virtual_attribute` on to the `ownership_mixin`, which is used by a few classes and makes it so that MiqExpressions will evaluate filter in SQL instead of pulling down all matching records to then interate over and filter in Ruby (causes a Massive N+1 on large datasets).  This requires a join for the `virtual_attribute` to function, so the `with_miq_group` scope does that with out pulling in any extra data, which is added in the controller which uses `MiqReport#paged_view_search` with an `MiqExpression`.


Metrics
-------
The data set for these metrics used a DB with approximately 20k `Vm` records.


**Before (Regular User)**

Started POST "/vm_or_template/tree_select/?id=:id"...
Completed 200 OK in 7242ms (Views: 2.0ms | ActiveRecord: 884.1ms)

|      ms |   bytes |   objects | queries | query (ms) |   rows |
|    ---: |    ---: |      ---: |    ---: |       ---: |   ---: |
| 6,043.1 |     N/A | 3,769,724 |      44 |    4,158.3 | 29,709 |


**After (Regular User)**

Started POST "/vm_or_template/tree_select/?id=:id"...
Completed 200 OK in 823ms (Views: 2.2ms | ActiveRecord: 46.6ms)

|      ms |   bytes |   objects | queries | query (ms) | rows |
|    ---: |    ---: |      ---: |    ---: |       ---: | ---: |
|   659.0 |     N/A |   816,802 |      30 |      131.9 |  397 |


**Before (Admin)**

Started POST "/vm_or_template/tree_select/?id=:id"...
Completed 200 OK in 94316ms (Views: 1.1ms | ActiveRecord: 13351.1ms)

|        ms | bytes |    objects | queries | query (ms) |    rows |
|      ---: |  ---: |       ---: |    ---: |       ---: |    ---: |
|  93,770.4 |   N/A | 57,852,016 |     112 |   78,603.0 | 349,084 |



**After (Admin)**

Started POST "/vm_or_template/tree_select/?id=:id"...
Completed 200 OK in 283ms (Views: 1.0ms | ActiveRecord: 12.6ms)

|      ms | bytes |   objects | queries | query (ms) | rows |
|    ---: |  ---: |      ---: |    ---: |       ---: | ---: |
|   190.2 |   N/A |    91,554 |      20 |       20.3 |   14 |


Additional Notes
----------------

This change drastically reduces the number of `rows` returns, time spend churning in ruby (sometimes to return nothing), decreases bandwidth and memory footprints, but the implementation in the controller is a bit hacky, and most likely requires some other changes now in other places to restore previous functionality.  It also attaches the `with_miq_group` scope to all other calls.  This should be fine for the most part, but could cause other regressions.  Not sure of a better way to limit the inclusion of this scope using the current mechanisms available without some other heavier refactoring required.


_Update:_

> It also attaches the `with_miq_group` scope to all other calls.

This is no longer the case as I have added controller methods to only add that scope when the MiqExpression "expression" includes the `owned_by_current_ldap_group` field.

NickLaMuro commented:
@miq-bot add_label performance, ui, refactoring

NickLaMuro commented:
FYI to the Reviewer:  Since the `owned_by_current_ldap_group` now requires `with_miq_groups` there are probably a lot more places where the controllers will need to be changed (or at least higher up the inheritance stack).  I skipped that for now incase code review ends up leading to a better implementation of the same SQL, and will fix before merging.

kbrock commented on app/models/mixins/ownership_mixin.rb:
```diff
@@ -11,9 +11,27 @@ module OwnershipMixin
+    virtual_attribute :owned_by_current_ldap_group, :boolean, :arel => (lambda do |t|
+      ldap_group = User.current_user.try(:ldap_group).to_s.downcase
+
+      t.grouping(MiqGroup.arel_table[:description].lower.eq(ldap_group))
```
could you use `arel_attribute`?
using that is why virtual attributes work at all

```ruby
t.grouping(MiqGroup.arel_attribute(:description).lower.eq(ldap_group))
```

Fryguy commented on app/models/mixins/ownership_mixin.rb:
```diff
@@ -11,9 +11,27 @@ module OwnershipMixin
     virtual_column :owning_ldap_group,                    :type => :string,     :uses => :miq_group
     virtual_column :owned_by_current_ldap_group,          :type => :boolean,    :uses => :owning_ldap_group
+
+    # Determine whether to return vms ownded by the current user or not
```
typo `ownded` => `owned`

NickLaMuro replied:
Yeah... need turn on my [spellchecker](https://github.com/NickLaMuro/dotfiles/blob/master/.vimrc.d/spellcheck.vim#L48-L49) on by default for ruby files too I guess 😞

Fryguy commented on app/models/mixins/ownership_mixin.rb:
```diff
@@ -11,9 +11,27 @@ module OwnershipMixin
+    # Will result in the following when used with MiqExpression:
+    #
+    #   WHERE ((LOWER("miq_groups"."description") = 'some_miq_group') = 'true')
+    virtual_attribute :owned_by_current_ldap_group, :boolean, :arel => (lambda do |t|
```
Is there no associated method?  (in other words, I didn't expect to see a bare virtual_attribute, but instead a virtual_column...is there a difference?)

Fryguy replied:
Should there be a unit test for this particular bit that it's working correctly?  The test below is more of an integration test (i.e. I don't see anything directly calling this method).

NickLaMuro replied:
@Fryguy Sorry, not shown in diff, but is at the bottom of the file.  Needed to be up here for application record to be loaded.

Fryguy replied:
Got it :+1:  Still wondering if this should be a virtual_column as opposed to virtual_attribute ( not sure what the difference is...@kbrock? )

NickLaMuro replied:
Technically equivalent in this case:

https://github.com/ManageIq/manageiq/blob/master/lib/extensions/ar_virtual.rb#L170-L180

I was referencing code in `app/models/hardware.rb` for doing this before I really understood how to harness `ar_virtual` for this, so this is mostly just imitation on my part.

kbrock replied:
@Fryguy these are the same

```ruby
virtual_attribute :name, :boolean, :arel => lambda...
virtual_column :name, :type => :boolean, :arel => lambda ...
```

Fryguy commented on app/models/mixins/ownership_mixin.rb:
```diff
@@ -11,9 +11,27 @@ module OwnershipMixin
 
   module ClassMethods
+    def with_miq_group
+      select(arel_table[Arel.star]).joins(:miq_group)
```
Is the `select *` necessary?  AR defaults to that regardless, so I'm not sure why it's needed unless you are trying to undo a previous select, in which I'm not sure this would work anyway.

NickLaMuro replied:
In this case, I am choosing **not** to return any data from the `miq_groups` table, and not from any other tables.  So it actually ends up being the following:

```SQL
SELECT "vms"."*"
```

Or whatever table this model actually happens to be associated with.

Fryguy commented on spec/models/mixins/ownership_mixin_spec.rb:
```diff
@@ -0,0 +1,26 @@
+        User.current_user = FactoryGirl.create :user,
+                                               :current_group => @miq_group,
+                                               :miq_groups => [@miq_group]
+        expect(User).to receive(:server_timezone).and_return("UTC")
```
I think there are helper methods to set up the current user with a group, but I could be wrong (@chrisarcand @jrafanie do you know?)

NickLaMuro replied:
I would love it if that was the case, and hope that my grep'ing skills were just sub par in this case.

chrisarcand replied:
`create :user_with_group`

NickLaMuro replied:
😍

chrisarcand replied:
BTW, prefer more helpers than instance vars in a before block.

jrafanie replied:
I think you want [stub_user](https://github.com/ManageIQ/manageiq/blob/a8635229337db4fb077295a98da003258969da10/spec/support/auth_helper.rb#L14) since that creates the user, group, stubs current_user, does the `server_timezone` thing.  It returns the user...so you can do `user.miq_group_id` on line 15.

Note, the user_with_group factory doesn't do the timezone stubbing above, AFAIK.

jrafanie replied:
Note, `stub_user` just landed and won't be backported.

NickLaMuro replied:
Thanks @jrafanie.  First looking into why my tests started failing all of the sudden, but I will see if that helps me out (or at least gives me a pattern to work with).

kbrock replied:
use:
```ruby
login_as_user FactoryGirl.create(:user_with_group)
```

NickLaMuro replied:
Few things:

* It is `login_as`, not `login_as_user`
* Not going to use `stub_user` because backwards compatible, and I would like this to go into `darga`.
* Logging in the user also need to happen **AFTER** the groups are created.  Found that out when I switched to @Fryguy 's [suggestion](https://github.com/ManageIQ/manageiq/pull/10290#discussion_r73751496) and the specs failed (a better way to test by the way since I had a false positive passing test).  `set_ownership` (I think) is called somewhere and assigns it to the `current_user`'s group on create (lost about 2 hours on this one).
* Also, unless this is a [view](https://github.com/ManageIQ/manageiq/blob/master/spec/spec_helper.rb#L48), [controller](https://github.com/ManageIQ/manageiq/blob/master/spec/spec_helper.rb#L54), or [helper](https://github.com/ManageIQ/manageiq/blob/master/spec/spec_helper.rb#L73) spec, both `stub_user` and `login_as` methods are not available in a model test.

Which raise another question, is there a better place for this test?  I really didn't want to do this in a controller test, but I guess I could.  Just seemed like it was a bit verbose to test this with a full action and rendering when I just wanted to know the results from the filter.

Thoughts?

chrisarcand replied:
@jrafanie Perhaps `stub_user` should be backported as the other macro and tests are pretty much broken behavior.

chrisarcand replied:
@NickLaMuro Yeah, I forgot to comment on this on Friday - those macros are controller specs.

jrafanie replied:
> @jrafanie Perhaps stub_user should be backported as the other macro and tests are pretty much broken behavior

@chrisarcand Maybe.  I can also see it being useful for model tests.  For this PR, I'd suggest @NickLaMuro do the stubbing/mocking he needs to do based on what `stub_user` does now and we can always backport and open up the auth helper for models when we know more.

chrisarcand replied:
:+1: @NickLaMuro [This](https://github.com/ManageIQ/manageiq/blob/master/spec/support/auth_helper.rb#L31) may or may not be the only line you need.

Fryguy commented on spec/models/mixins/ownership_mixin_spec.rb:
```diff
@@ -0,0 +1,26 @@
+      it "returns results only part of the miq_group" do
+        vm_ids = report.paged_view_search(search_opts).first.map(&:id)
+        vms    = Vm.where(:id => vm_ids)
+        expect(vms.all? {|vm| vm.miq_group_id == @miq_group.id }).to be true
```
Can you add a test for vms that are in a group but not *this* particular group?

Additionally, from a test structure perspective, I would expect that the vm_with_group would be in the test itself since it links the setup to the expectation and makes it clearer to me what the result should look like...something like

```ruby
expect(vms).to match_array [vm_with_group]
```

NickLaMuro replied:
Yeah, wasn't sure if we were starting from a clean state (no records in the `vms` table), so I was mostly working around that by making sure the items returned only are part of the `MiqGroup`, and not ones created without.

This is silly now that I think about it because I am making the `MiqGroup` as part of this test, so only the `vms` that I am creating here should be include... I will make that change.

For the record, I think this test is kind of a hack in it of itself, but I was mostly doing it as a way of confirming existing functionality, and that I didn't break it when adding the new additions (which technically I did because `with_miq_group` is now a required scope).

NickLaMuro commented:
@Fryguy I apologize for the delay on getting the updates to this PR out, but the fixes you requested should be here now.

Some items of note:

* Changed the test in the first commit to be a `shared_example` instead of a mixin test, since what we care about is that the models that inherit this code behave the same.  This triples the amount of time this is run, since I include it in three model tests that include this module, but it will make sure the same scenarios are run in each and all three work. ( @chrisarcand and/or @jrafanie might have some opinions on this)
* I am not using any of the `User` helpers for Reasons™.  I can go in more depth if requested, but basically they cause more trouble (and code) then it is worth.
* Added unit tests that confirm the functionality of the attribute (in the second commit), so that could use a look.
* As noted there and in a commit message, the functionality of `owned_by_current_ldap_group filter` is somewhat counter intuitive.  When requesting it with "true" (meaning "I want the objects that are part of my ldap group"), it works as expected, but when passing false, it will only return objects that have a group, not "all other objects that aren't a part of your LDAP group".  This seems a bit weird, but I didn't fix it because the original logic that confirms this has been there a while, and changing both the ruby method and the virt attribute seemed like the wrong choice (we can update this later if we feel it is the wrong logic).
* Updated the rest of the controllers that I was able to find that would need updated to support this change (that I was able to confirm used this).  This was rebased into the third commit.


### EXTRA INFO (don't read if you don't care)

Expanding on the current logic of `owned_by_current_ldap_group`, this is what would need to change to support the logic **I think** is intuitive:

```diff
diff --git a/app/models/mixins/ownership_mixin.rb b/app/models/mixins/ownership_mixin.rb
index 90708fa..abc38da 100644
--- a/app/models/mixins/ownership_mixin.rb
+++ b/app/models/mixins/ownership_mixin.rb
@@ -13,23 +13,41 @@ module OwnershipMixin
     virtual_column :owned_by_current_ldap_group,          :type => :boolean,    :uses => :owning_ldap_group
 
     # Determine whether to return vms owned by the current user or not
-    # Requires `with_miq_group` scope to function
+    # Requires `with_miq_group` scope to function, and the COALESCE is there
+    # incase a miq_group doesn't exist for the model (since we are doing a left
+    # join in `with_miq_group`)
     #
-    # Resulting SQL: "(LOWER(miq_groups.description) = 'some_miq_group'"
+    # Resulting SQL: "(COALESCE(LOWER("miq_groups"."description"), '')) = 'some_miq_group')"
     #
     # Will result in the following when used with MiqExpression:
     #
-    #   WHERE ((LOWER("miq_groups"."description") = 'some_miq_group') = 'true')
+    #   WHERE ((COALESCE(LOWER("miq_groups"."description"), '') = 'some_miq_group') = 'true')
+    #
     virtual_attribute :owned_by_current_ldap_group, :boolean, :arel => (lambda do |t|
       ldap_group = User.current_user.try(:ldap_group).to_s.downcase
 
-      t.grouping(MiqGroup.arel_table[:description].lower.eq(ldap_group))
+      empty_string = Arel::Nodes::SqlLiteral.new "''"
+      description  = MiqGroup.arel_table[:description].lower
+      coalesce     = Arel::Nodes::NamedFunction.new "COALESCE", [description, empty_string]
+      t.grouping(coalesce.eq(ldap_group))
     end)
   end
 
   module ClassMethods
+
+    # Does a left join on the miq_groups table, but only selects from the
+    # current table.  Generated sql should look something like this:
+    #
+    #   SELECT vms.*
+    #   FROM [ownership_table]
+    #   LEFT OUTER JOIN miq_groups ON [ownership_table].miq_group_id = miq_group.id
+    #
     def with_miq_group
-      select(arel_table[Arel.star]).joins(:miq_group)
+      miq_group_table  = MiqGroup.arel_table
+      group_left_join  = arel_table.join(miq_group_table, Arel::Nodes::OuterJoin)
+                           .on(arel_table[:miq_group_id].eq(miq_group_table[:id]))
+                           .join_sources
+      select(arel_table[Arel.star]).joins(group_left_join)
     end
 
     def set_ownership(ids, options)
@@ -92,7 +110,6 @@ module OwnershipMixin
   end
 
   def owned_by_current_ldap_group
-    ldap_group = User.current_user.try(:ldap_group)
-    ldap_group && owning_ldap_group && (owning_ldap_group.downcase == ldap_group.downcase)
+    owning_ldap_group.to_s.downcase == ldap_group.to_s.downcase
   end
 end
```

The above isn't fool proof, so I don't want to add it, but I thought I would bring it to people's attention incase it was actually unintended behavior (and to justify (slightly) why I was taking so long with the tests...).

NickLaMuro commented:
^ Go home `rubocop`, you are drunk

_These are all a `WONTFIX` since I didn't do much more than add a single line to this method.  We might want to tweak how the bot handles this, since I wasn't getting these errors when using `rubocop-git` against master (different than what we do, I realize, but it is the closest/simplest way to replicate MiqBot locally), but that is something to discuss somewhere else._

NickLaMuro commented:
There has been even some more churn on this PR since https://github.com/ManageIQ/manageiq/pull/10290#issuecomment-239031174 ... sigh...

* Moved the `ownership_mixin` tests to a `shared_example` so they can be reused in the models that use them
  - This means these tests are now repeated 3 times for each model
  - Added an alias in the spec helper to make the `include` for the shared_example read a little better
  - Removes duplication, but definitely makes it a bit harder to grok
* `set_named_scope` is now MUCH more complex.  It is basically needs to evaluate any `MiqExpression` that might be set, and a MiqExpression can have multiple fields it is searching on (unfortunately).
  - This makes use of `MiqExpression::get_cols_from_expression` in `miq_search_exp_fields`, which now returns an array of fields.
  - This is unfortunately a bit slower than previous iterations, but not really noticeable since this method is only run once.
  - There have been a bunch of iterations on these controller methods, that are not shown because of rebases...
* `set_named_scope` is added as a `before_action` only for the explorer route.  Because of how `tree_select` sets the `@edit` and `session[:edit]` within the action, this was require to add this line in the method instead of as a `before_action`
* `set_named_scope` has a `case` statement now.
  - This is first to limit setting that scope to only when we need it for the MiqExpression, not all the time
  - This also sets up for a future PR where I will need to use a different scope for a different filter.
  - The `proc` + `includes?` is kinda smelly, but I couldn't think of a more terse way of representing that.  Suggestions welcome.

jrafanie commented on app/models/mixins/ownership_mixin.rb:
```diff
@@ -11,9 +11,27 @@ module OwnershipMixin
+    #
+    #   WHERE ((LOWER("miq_groups"."description") = 'some_miq_group') = 'true')
+    virtual_attribute :owned_by_current_ldap_group, :boolean, :arel => (lambda do |t|
+      ldap_group = User.current_user.try(:ldap_group).to_s.downcase
```
@chrisarcand 😭 We might have to look at this type of thing with multiple entitlements.

kbrock commented on app/controllers/mixins/vm_show_mixin.rb:
```diff
@@ -70,6 +70,24 @@ def identify_vm
+  def set_named_scope
+    @named_scope =
+      case miq_search_exp_fields
+      when ->(fields) { fields.include? "owned_by_current_ldap_group" } then :with_miq_group
```
nitpick (aka you can ignore)

could you just use an if instead?

```ruby
@named_scope = miq_search_exp_fields.include? "owned_by_current_ldap_group" ? :with_miq_group : nil
```

NickLaMuro replied:
From https://github.com/ManageIQ/manageiq/pull/10290#issuecomment-239876947

> * set_named_scope has a case statement now.
>   - This is first to limit setting that scope to only when we need it for the MiqExpression, not all the time
>   - **This also sets up for a future PR where I will need to use a different scope for a different filter.**
>   - The proc + includes? is kinda smelly, but I couldn't think of a more terse way of representing that. Suggestions welcome.

I eventually need to support other scopes, hence the case statement, not a ternary.  That said, an `if` might work better at this point:

```ruby
@named_scope = 
  if miq_search_exp_fields.include? "owned_by_current_ldap_group"
    :with_miq_group
  elsif miq_search_exp_fields.include? "owned_by_current_user"
    :with_evm_owner # This scope to be added in a future PR
  end
```

V.S. what it will be:

```ruby
@named_scope =
  case miq_search_exp_fields
  when ->(f) { f.include? "owned_by_current_ldap_group" } then :with_miq_group
  when ->(f) { f.include? "owned_by_current_user" } then :with_evm_owner
  end
```

chrisarcand commented on spec/spec_helper.rb:
```diff
@@ -28,6 +28,8 @@
     c.syntax = :expect
   end
 
+  config.alias_it_should_behave_like_to :includes_mixin, 'includes mixin:'
```
👎 

`include_examples` and/or `it_behaves_like` are already used throughout our specs and adding our own customized and undocumented alias seems like a bad idea to me. I see where it kinda sorta sounds better given the mixin happy nature of our code but I think people would be much happier knowing exactly what's going on right away (or at least Googleable) with `include_examples "miq ownership"`

NickLaMuro replied:
I am honestly not in the mood for fighting this, so whatever, I will change it.

But just for the record, here is why I didn't use `include_examples` from the get go.  From [the rspec docs](http://www.relishapp.com/rspec/rspec-core/v/3-5/docs/example-groups/shared-examples):

```
include_examples "name"      # include the examples in the current context
it_behaves_like "name"       # include the examples in a nested context
it_should_behave_like "name" # include the examples in a nested context
matching metadata            # include the examples in the current context
```

Since I used `before(:context)`/`after(:context)` in the top level here, using `include_examples` will break other tests this is include with if they rely on a clean DB state to do there tests (which at least one of them does).

`it_behaves_like "miq ownership"` sounds like butt...

kbrock commented:
10685 was merged instead
